hathor/indexes/manager.py

```diff
@@ -133,9 +138,12 @@ def del_tx(self, tx: BaseTransaction, *, remove_all: bool = False, relax_assert:
         if self.tokens:
             self.tokens.del_tx(tx)
 
+        self.deps.del_tx(tx)
```

Comment:
Is this idempotent as well?

Reply:
While I'm not sure (I'll check it in a bit), I don't believe wider `IndexesManager.del_tx` is meant to be idempotent. So I don't think there should be any problem.

Reply:
The implementations seem to be in fact idempotent, but I also don't think it's intentional, they just happen to be idempotent.

But also taking a better look again into the indexes as a whole, it looks like `IndexesManager.del_tx` _is_ meant to be _sort of_ idempotent, its just to clear at all. I think we could definitely improve the documentation of the indexes and how we expect everything to work.

Right now event the word "idempotent" is only present in the documentation of some of the methods involved in this PR. I think this might not give a clear view to someone new to the code, it might give the impression that only this index in particular require that property, while actually most, if not all, of `add/del` methods from the indexes need it as well.

From what I understand it basically boils down to a transaction being able to be "removed twice" or "added twice" to the indexes. It can _technically_ be "removed twice" when first becoming voided and second being completely removed , although there is no codepath that leads to a transaction being "completely removed". And it can also _techincally_ be "added twice" when first being newly received and second when being unvoided, although that implies it was voided when it was received (which I don't think is possible, I think we add as if it wasn't voided, then run the consensus, which can then void it), however it could also be: 1) received, 2) voided, 3) unvoided, and this would map to: 1) add 2) remove *not all* 3) add, if we didn't have the distinction between "remove *all*" and "remove *not all*", then there wouldn't be a requirement for an "idempotant add" on some indexes, but because we do, at least the `all_tips` and `sorted_all` indexes need an "idempotent add" (since their "remove" will not be called on "remove *not all*").

Anyway, I don't think this is something special to the dependencies indexes, and I'd think it's best to have a front to improve the documentation and testing of the whole indexing mechanism.

Reply:
Makes a lot of sense. And I agree that we haven't clearly discussed and documented which methods are idempotent. I feel we should create a new issue to do it properly in the future.

Reply:
Done: https://github.com/HathorNetwork/hathor-core/issues/409

Reply:
Thanks!